pyomo.contrib.alternative_solutions

whart222 opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

## Summary/Motivation:

A new contrib package for generating alternative solutions

## Changes proposed in this PR:

- Adding pyomo.contrib.alternative_solutions

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

whart222 commented:
I have one generate concern about this PR, which is the fact that the appsi- and non-appsi- solver interfaces are different enough that there are many explicit checks for the solver type with different logic.  I know that there have been efforts to standardize the solver interfaces, so perhaps we can resolve some of these discrepancies?  I tried without luck a month or so ago.

emma58 commented:
@whart222, if you're ready for me to start looking at this would you run `black`? Then the rest of the tests will run. Thank you!

whart222 commented:
@jsiirola @emma58 I'm not sure what is going on here.  I merged-in updates from the main branch, and now I'm seeing weird test failures. I *think* that I've resolved all of the Git test issues.

jsiirola commented:
A quick look and what I see is that an update to the spell checker action (part of the linter test) is catching new spelling errors.  The fixes for that are in #3303, which should get merged after it is reviewed sometime this morning (then we will need to rebuild the entire backlog of PRs ... which will take all day).

blnicho commented:
@whart222 looks like there are a few real test failure on Windows and there are some unguarded numpy imports causing the slim build to fail.

whart222 commented:
@blnicho Thanks for the heads-up.  It took me a hot minute to see that I could scroll through the tests.  Fixing now.

whart222 commented:
I've removed the "no_time" tests, which confirm expected behavior when the timelimit is set "too small".  However, on several of the Windows test platforms these resulted in a normal completion to an optimal solution.  Perhaps the test problems are too easy and the time granularity on Windows machines is large?  I'm not sure, but I don't see an effective way to diagnose this issue.

whart222 commented:
@emma58 @jsiirola @blnicho @mrmundt Are these test failures indicative of a broader issue with codecov?

mrmundt commented:
@whart222 - The problem is in the way test results are asserted. `assert` will raise errors when the assertion is invalid, whereas `unittest`'s `self.assert_____` (insert appropriate assertion here) allows tests to run and records a pass or fail. `pytest` assumes the usage of the latter, not the former.

jsiirola commented:
> @whart222 - The problem is in the way test results are asserted. 

Actually, the root cause is more nefarious: pytest can usually catch and proceed when tests fail with `assert` (although @mrmundt is correct: in Pyomo we strongly prefer that tests use the unittest `self.assert___` methods).  The root problem is that test file is running the tests *at the module scope* and not inside a `TestCase` class: so the assertion is failing on _import_ (when pytest is attempting to collect all the tests in the project), which cases pytest to die.

whart222 commented:
@jsiirola @mrmundt Well, I didn't realize that a commit had been made that included a test failure.  I've commented that out for now until we can confirm whether/how that test is accurate.  This is for new functionality that is being added, so I'm inclined to proceed with the PR and than do an update for that later.

mrmundt commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,235 @@
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+try:
+    from gurobipy import GRB
+
+    gurobi_available = True
+except:
+    gurobi_available = False
```
It is recommended to do this logic via `pyomo.common.dependencies`:

```
from pyomo.common.dependencies import gurobipy, gurobipy_available
if gurobipy_available:
   ... extra logic ...
```

mrmundt commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,235 @@
+                    if continuous_var[var].value > self.zero_threshold:
+                        num_non_zero += 1
+                        non_zero_basic_expr += binary_var[var]
+            # TODO: JLG - If we want to add the mixed binary case, I think we
+            # need to do it here. Essentially we would want to continue to
+            # build up the num_non_zero as follows
+            # for binary in binary_vars:
+            # if binary.value > 0.5:
+            # num_non_zero += 1 - binary
+            # else:
+            # num_non_zero += binary
```
Needed?

whart222 replied:
Same as earlier.  These comments refer to features we haven't flushed out yet.

mrmundt commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -0,0 +1,115 @@
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+try:
+    import gurobipy
+
+    gurobi_available = True
+except:
+    gurobi_available = False
```
It is recommended to do this logic via `pyomo.common.dependencies`:

```
from pyomo.common.dependencies import gurobipy, gurobipy_available
if gurobipy_available:
   ... extra logic ...
```

mrmundt commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -0,0 +1,147 @@
+        A map between Pyomo variables and their values for a solution.
+    fixed_vars : ComponentSet
+        The set of Pyomo variables that are fixed in a solution.
+    objectives : ComponentMap
```
Nitpicking - should be `objective`?

whart222 replied:
I don't think so.  A model may have more than one Objective component specified.

mrmundt replied:
The var in the code is `objective` - so something doesn't line up! ;)

whart222 replied:
You're right.  The code isn't as general as I had thought.  {sigh}

I fixed the docs.

mrmundt commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -0,0 +1,147 @@
+        Get a dictionary of variable name-variable value pairs.
+    get_fixed_variable_names(self):
+        Get a list of fixed-variable names.
+    def get_objective_name_values(self):
```
```suggestion
    get_objective_name_values(self):
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_aos_utils.py:
```diff
@@ -0,0 +1,292 @@
+import pytest
```
Please add the copyright statement:

```
#  ___________________________________________________________________________
#
#  Pyomo: Python Optimization Modeling Objects
#  Copyright (c) 2008-2024
#  National Technology and Engineering Solutions of Sandia, LLC
#  Under the terms of Contract DE-NA0003525 with National Technology and
#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
#  rights in this software.
#  This software is distributed under the 3-clause BSD License.
#  ___________________________________________________________________________
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_aos_utils.py:
```diff
@@ -0,0 +1,292 @@
+import pytest
+
```
Please instead use:
```
from pyomo.common import unittest as unittest
```
instead of `pytest`. `pytest` is accessible via: `unittest.pytest`

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_aos_utils.py:
```diff
@@ -0,0 +1,292 @@
+import pytest
+
+try:
+    from numpy.linalg import norm
+
+    numpy_available = True
+except:
+    numpy_available = False
```
It is recommended to do this logic via `pyomo.common.dependencies`:

```
from pyomo.common.dependencies import numpy as np, numpy_available
if numpy_available:
   ... extra logic ...
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,137 @@
+import pytest
```
Please add the copyright statement:

```
#  ___________________________________________________________________________
#
#  Pyomo: Python Optimization Modeling Objects
#  Copyright (c) 2008-2024
#  National Technology and Engineering Solutions of Sandia, LLC
#  Under the terms of Contract DE-NA0003525 with National Technology and
#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
#  rights in this software.
#  This software is distributed under the 3-clause BSD License.
#  ___________________________________________________________________________
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,137 @@
+import pytest
+
+try:
+    from numpy.testing import assert_array_almost_equal
+
+    numpy_available = True
+except:
+    numpy_available = False
```
It is recommended to do this logic via `pyomo.common.dependencies`:

```
from pyomo.common.dependencies import numpy as np, numpy_available
if numpy_available:
   ... extra logic ...
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,137 @@
+import pytest
+
```
Please instead use:
```
from pyomo.common import unittest as unittest
```
instead of `pytest`. `pytest` is accessible via: `unittest.pytest`

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_cases.py:
```diff
@@ -0,0 +1,430 @@
+from itertools import product
```
Rather than repeating even more times... Please add the copyright statement to all new files:

```
#  ___________________________________________________________________________
#
#  Pyomo: Python Optimization Modeling Objects
#  Copyright (c) 2008-2024
#  National Technology and Engineering Solutions of Sandia, LLC
#  Under the terms of Contract DE-NA0003525 with National Technology and
#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
#  rights in this software.
#  This software is distributed under the 3-clause BSD License.
#  ___________________________________________________________________________
```

mrmundt reviewed (changes requested):
Specifically in the `test_*` files:

- Please add the copyright statement:

```
#  ___________________________________________________________________________
#
#  Pyomo: Python Optimization Modeling Objects
#  Copyright (c) 2008-2024
#  National Technology and Engineering Solutions of Sandia, LLC
#  Under the terms of Contract DE-NA0003525 with National Technology and
#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
#  rights in this software.
#  This software is distributed under the 3-clause BSD License.
#  ___________________________________________________________________________
```

- Please make sure you are using `pyomo.common.unittest` and not directly importing `pytest` (because it's available in `pyomo.common.unittest` already).
- Use the nifty `pyomo.common.dependencies` for things like `numpy` and `gurobipy`

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+        include_integer : boolean
+            Boolean indicating that integer variables should be included.
+        include_fixed : boolean
+            Boolean indicating that fixed variables should be included.
```
Ty.  Done.

whart222 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,246 @@
+    for parameter, value in solver_options.items():
+        opt.options[parameter] = value
+    #
+    # Appsi-specific configurations
+    #
+    use_appsi = False
```
I tried using contrib.solver without luck.  I vote that we wrap-up this PR as-is, and then revisit the solver interfaces together.

mrmundt replied:
Good note!

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,351 @@
+
+def enumerate_linear_solutions(
+    model,
+    *,
```
Same answer. ;)

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,351 @@
+
+    if variables == "all":
+        all_variables = aos_utils.get_model_variables(model, "all")
+    # else:
+    #     binary_variables = ComponentSet()
+    #     non_binary_variables = []
+    #     for var in variables:
+    #         if var.is_binary():
+    #             binary_variables.append(var)
+    #         else:
+    #             non_binary_variables.append(var.name)
+    #     if len(non_binary_variables) > 0:
+    #         print(('Warning: The following non-binary variables were included'
+    #                'in the variable list and will be ignored:'))
+    #         print(", ".join(non_binary_variables))
+    # all_variables = aos_utils.get_model_variables(model, 'all',
+    #                                               include_fixed=True)
```
We're not sure.  This refers to functionality that we'd like to add, but which it's not quite clear if it will work as expected.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,351 @@
+    for var in all_variables:
+        assert var.is_continuous(), "Model must be an LP"
+
+    use_appsi = False
+    # TODO Check all this once implemented
+    if "appsi" in solver:
+        use_appsi = True
```
Same as above.  The solver interfaces need to be revisited in this code.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,351 @@
+
+        if debug:
+            model.pprint()
+            # print("Writing test{}.lp".format(solution_number))
+            # cb.write("test{}.lp".format(solution_number))
```
Nope.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,351 @@
+                for var in continuous_var:
+                    if continuous_var[var].value > zero_threshold:
+                        num_non_zero += 1
+                        # WEH - I don't think you need to add the binary variable.  It
+                        #       should be automatically added when used.
+                        # if var not in binary_var:
+                        #    binary_var[var]
```
I'm not sure.  This is a comment from my initial code review.  There are a bunch of comments like this that we need to consider to verify correctness of the code.

I *did* add a lot of tests, and the code *seems* to work.

emma58 replied:
This comment could be removed.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,351 @@
+                cb.basic_last_slack,
+            ]
+
+            num_non_zero = 0
+            force_out_expr = -1
+            non_zero_basic_expr = 1
```
I added comments.

whart222 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,361 @@
+
+def obbt_analysis(
+    model,
+    *,
```
Same answer.

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -0,0 +1,115 @@
+
+def gurobi_generate_solutions(
+    model,
+    *,
```
Same answer.

whart222 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -0,0 +1,115 @@
+#  ___________________________________________________________________________
```
The issue was test marking.  Thanks for the help!

whart222 commented:
@mrmundt I think I've addressed the concerns, but I haven't closed out ones that we might need to discuss further.

I'm strongly inclined to approve the PR as-is and then address the (1) missing AOS solver features, and (2) inconsistencies between solver APIs.  I have FY25 funding to continue developing/refining these capabilities.

mrmundt commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,246 @@
+
+def enumerate_binary_solutions(
+    model,
+    *,
```
I don't have strong feelings about this - the question mainly came up because of user experience. If you provide defaults, then the user does not have to put in everything.

whart222 replied:
That's still true with this syntax.  With defaults, you don't need to specify the arguments after the "*".  But any arguments you do specify must be with a keyword.  For example

`enumerate_binary_solutions(model, quiet=True)`

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_aos_utils.py:
```diff
@@ -0,0 +1,300 @@
+        self.assertEqual(None, cons[1].upper)
+        self.assertEqual(9, cons[1].lower)
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+                m, num_solutions=100, solver=mip_solver, solver_options={"TimeLimit": 0}
+            )
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+        unique_solns_by_obj = [val for val in Counter(objectives).values()]
+        assert_array_almost_equal(unique_solns_by_obj, m.num_ranked_solns)
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+        unique_solns_by_obj = [val for val in Counter(objectives).values()]
+        assert_array_almost_equal(unique_solns_by_obj, [1, 1, 1, 1])
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+        )
+        assert_array_almost_equal(objectives, m.ranked_solution_values[:3])
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+        )
+        assert_array_almost_equal(objectives, [6, 3, 1])
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_lp_enum.py:
```diff
@@ -0,0 +1,100 @@
+        assert sols[0].objective_value == unittest.pytest.approx(6.789473684210527)
+        assert sols[1].objective_value == unittest.pytest.approx(3.6923076923076916)
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_lp_enum.py:
```diff
@@ -0,0 +1,100 @@
+            print(s)
+        assert len(sols) == 6
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+@unittest.pytest.mark.default
+class TestOBBTUnit:
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        with unittest.pytest.raises(AssertionError):
+            obbt_analysis_bounds_and_solutions(m, variables=[m.x], solver=mip_solver)
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        for var, bounds in all_bounds.items():
+            assert_array_almost_equal(bounds, m.continuous_bounds[var])
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        for var, bounds in all_bounds.items():
+            assert_array_almost_equal(bounds, m.continuous_bounds[var])
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        assert len(solns) == 2 * len(all_bounds) + 1
+        assert m._obbt.optimality_tol_rel.lb == unittest.pytest.approx(2.5)
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        assert len(solns) == 2 * len(all_bounds) + 1
+        assert m._obbt.optimality_tol_abs.lb == unittest.pytest.approx(3.01)
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        for var, bounds in all_bounds.items():
+            assert_array_almost_equal(bounds, m.continuous_bounds[var])
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        assert bounds_tightened
+        assert bounds_not_tightened
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+            assert_array_almost_equal(bounds, m.continuous_bounds[var])
+        assert len(solns) == num
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_shifted_lp.py:
```diff
@@ -0,0 +1,68 @@
+@unittest.pytest.mark.default
+class TestShiftedIP:
+
+    @unittest.pytest.mark.skipif(not numpy_available, reason="Numpy not installed")
```
```suggestion
    @unittest.skipIf(not numpy_available, "Numpy not installed")
```

mrmundt commented on pyomo/contrib/alternative_solutions/tests/test_solnpool.py:
```diff
@@ -0,0 +1,149 @@
+
+
+@unittest.skipUnless(gurobipy_available, "Gurobi MIP solver not available")
+# @unittest.pytest.mark.skipif(not gurobipy_available, reason="Gurobi MIP solver not available")
```
Remove

mrmundt reviewed (changes requested):
Some residual pytest.mark.skip labels that will likely cause issues.

whart222 commented:
@mrmundt  I've updated the tests.  We can see if the tests pass now.

whart222 commented:
@mrmundt I'm having trouble finding the unresolved change request.  Can you help?

blnicho commented:
> @mrmundt I'm having trouble finding the unresolved change request. Can you help?

@whart222 I think it was just an artifact of GitHub's review system. I dismissed Miranda's (now stale) review and re-requested a review from her to clear the change request.

whart222 commented:
@emma58 @mrmundt @blnicho Are there any outstanding issues that need to be resolved?  I think I've resolved all of the review concerns.

emma58 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,370 @@
+            results = opt.solve(
+                model, warmstart=warmstart, tee=tee, load_solutions=False
+            )
+        except:
+            # Assume that we failed b.c. of warm starts
+            results = None
```
I think the bare `except` here is probably pretty dangerous. Can you specifically catch the warmstart exception and if it's not that, just raise it to be safe?

whart222 replied:
I don't disagree.  But can we trust that solvers will consistently return the same error for failures during a warmstart?  Do we have such an exception defined?

whart222 replied:
@emma58 Do you have a recommendation for how to handling the exception handling differently?

emma58 replied:
I'm actually unclear on when you get an exception because of warmstarts... At least with Gurobi it usually just rejects your warmstart and goes forward. So honestly if it were me I would not catch anything here and assume all exceptions are something the user needs to know. But maybe @jlgearh knows more--I'm not sure what the original reason behind this try-except was?

whart222 replied:
I cleaned this up.  This is catching the case where a solver doesn't implement the 'warmstart' option.

emma58 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,370 @@
+            if idx == 0:
+                variable_bounds[var] = [None, None]
+
+            if hasattr(aos_block, "var_objective"):
+                aos_block.del_component("var_objective")
+
+            aos_block.var_objective = pe.Objective(expr=var, sense=sense)
```
Pyomo will let you assign to `var_objective.expr` and `var_objective.sense`, so you don't have to add and remove the component every iteration if you don't want.

whart222 replied:
Really?  I thought Pyomo complained when you re-added a component to a model.  This logic avoids those warnings.

jsiirola replied:
It will complain if you replace a component with another component.  What @emma58 was suggesting was to just assign a new expression on the existing Objective.  That is, 
```python
aos_block.var_objective.expr = var
aos_block.var_objective.sense = sense
```

whart222 replied:
Done, but really it's not more parsimonious b.c. you need an if/else statement to cover first time you add the objective.

emma58 replied:
Since it's on the `aos_block`, you could just make the objective when you originally make the Block and then not need the if/else since you'll know it's already there.

emma58 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,370 @@
+
+                if refine_discrete_bounds and not var.is_continuous():
+                    if sense == pe.minimize and var.lb < obj_val:
+                        bound_name = var.name + "_" + str.lower(bound_dir)
```
Are you sure this name will be unique? Also, why not just change the `bounds` on the `Var` object?

whart222 replied:
We are only changing data in the AOS block to ensure that the model values don't get changed.

Within that block, this name will be unique if the variable names are unique.  I'm not sure that this will correctly handle variables on different blocks with the same name.  @jlgearh ???

emma58 replied:
We discussed this and will switch to using the unique name function or changing the bounds and restoring them at the end.

emma58 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,370 @@
+
+            iteration += 1
+
+    # TODO - Remove this block
+    aos_block.deactivate()
+    orig_objective.activate()
```
Does this TODO need doing?

whart222 replied:
@jlgearh ???  I'm not sure why you don't just delete aos_block from the model.

emma58 replied:
We will leave the Block (deactivated), so this TODO can be removed.

whart222 replied:
Done

emma58 commented on pyomo/contrib/alternative_solutions/shifted_lp.py:
```diff
@@ -0,0 +1,225 @@
+    slack_var.setub(slack_ub)
+
+
+def get_shifted_linear_model(model, block=None):
+    """
+    Converts an (MI)LP with bounded (discrete and) continuous variables
+    (l <= x <= u) into a standard form where where all continuous variables
+    are non-negative reals and all constraints are equalities. For a pure LP of
+    the form,
+
+    min/max cx
+    s.t.
+        A_1 * x  =  b_1
+        A_2 * x <= b_2
+        l <= x <= u
+
+    a problem of the form,
+
+    min/max c'z
+    s.t.
+        Bz = q
+        z >= 0
```
Not necessary for this PR, but in the future this could be modified to use the `compile_standard_form` writer from `repn.plugins.standard_form`.

whart222 replied:
Good idea!

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -0,0 +1,115 @@
+        Solver option-value pairs to be passed to the Gurobi solver.
+    tee : boolean
+        Boolean indicating that the solver output should be displayed.
+    quiet : boolean
+        Boolean indicating whether to suppress all output.
```
Another place it would be good to switch to a logger.

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -0,0 +1,115 @@
+    #
+    # Setup gurobi
+    #
+    if not gurobipy_available:
+        return []
```
Same question as above here--is this safe behavior or should this throw an error?

whart222 replied:
I think this is the desirable behavior.

emma58 commented on pyomo/contrib/alternative_solutions/solnpool.py:
```diff
@@ -0,0 +1,115 @@
+    if not gurobipy_available:
+        return []
+    opt = appsi.solvers.Gurobi()
+    if not opt.available():  # pragma: no cover
+        return []
```
And same here? This should probably throw an error saying that appsi needs to be installed, yes?

whart222 replied:
See above.  We're being consistent in the current API.

emma58 commented on pyomo/contrib/alternative_solutions/solution.py:
```diff
@@ -0,0 +1,158 @@
+            self.to_string(
+                round_discrete=round_discrete, sort_keys=sort_keys, indent=indent
+            )
+        )  # pragma: no cover
```
This should be hard to test, should it? Or honestly leaving it uncovered for now is probably fine.

whart222 replied:
Capturing I/O can be tricky.

jsiirola replied:
```python
with pyomo.common.tee.capture_output() as OUT:
   # run test
self.assertEqual(OUT.getvalue(), reference_string)
```
If you have issues with the thing you are testing printing directly to the underlying C file descriptors (some solvers do this), use `capture_output(capture_fd=True)`

emma58 commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+        assert len(results) == 1
+        assert results[0].objective_value == unittest.pytest.approx(5)
+
+    def Xtest_no_time(self, mip_solver):
```
Maybe you could skip the test with a comment that it's fragile? I'm worried with this solution we'll forget about it entirely.

whart222 replied:
Done.

emma58 commented on pyomo/contrib/alternative_solutions/tests/test_lp_enum.py:
```diff
@@ -0,0 +1,100 @@
+@unittest.pytest.mark.default
+class TestLPEnum:
+
+    def Xtest_no_time(self, mip_solver):
```
Same comment here--it might be good to skip it and leave a breadcrumb about fragility.

whart222 replied:
Done

emma58 commented on pyomo/contrib/alternative_solutions/tests/test_balas.py:
```diff
@@ -0,0 +1,145 @@
+pytestmark = unittest.pytest.mark.parametrize("mip_solver", solvers)
+
+
+@unittest.pytest.mark.default
+class TestBalasUnit:
```
Generally we do this like it is in `test_aos_utils.py` where we inherit from `unittest.TestCase` and then you can use the different types of assertion methods that come on that base class rather than just `assert` statements. Not sure if it needs to hold up this PR, but it would be preferable to use that style since failures will be much easier to interpret. (I think this comment holds for many of the test files below as well.)

whart222 replied:
This won't be a simple change given the pytest logic I'm leveraging to parameterize tests.  Let's do this after the PR.

emma58 commented on pyomo/contrib/alternative_solutions/tests/test_obbt.py:
```diff
@@ -0,0 +1,214 @@
+        for var, bounds in all_bounds.items():
+            assert_array_almost_equal(bounds, m.var_bounds[var])
+
+    def Xtest_no_time(self, mip_solver):
```
Same comment as above here.

whart222 replied:
Done

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+    normalizing a standard multivariate Gaussian distribution.
+    """
+    global rng
+    iterations = 1000
+    min_norm = 1e-4
```
Change made.

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+
+    assert (
+        rel_opt_gap is None or rel_opt_gap >= 0.0
+    ), "rel_opt_gap must be None of >= 0.0"
```
Fixed

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+
+    active_objs = []
+    for o in model.component_data_objects(pe.Objective, active=True):
+        objs = o.values() if o.is_indexed() else (o,)
+        for obj in objs:
```
I don't see how to change the for loop.

jsiirola replied:
```diff
- for o in model.component_data_objects(pe.Objective, active=True):
- objs = o.values() if o.is_indexed() else (o,)
-     for obj in objs:
-         active_objs.append(obj)
+ active_objs = list(model.component_data_objects(pe.Objective, active=True))
```

whart222 replied:
Done

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+    ), "rel_opt_gap must be None of >= 0.0"
+    assert (
+        abs_opt_gap is None or abs_opt_gap >= 0.0
+    ), "abs_opt_gap must be None of >= 0.0"
```
Fixed

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+    global rng
+    iterations = 1000
+    min_norm = 1e-4
+    idx = 0
+    while idx < iterations:
+        samples = rng.normal(size=num_dimensions)
+        samples_norm = norm(samples)
+        if samples_norm > min_norm:
+            return samples / samples_norm
+        idx += 1  # pragma: no cover
```
Yeah, that's simpler.

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+        ----------
+        model : ConcreteModel
+            A concrete Pyomo model.
+        components: 'all' or a collection Pyomo components
```
Changing the default touched the APIs of all of the public functions in the module.  I'm hoping that's what you wanted.

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+                        include_integer,
+                        include_fixed,
+                    )
+            elif hasattr(comp, "ctype") and comp.ctype == pe.Var:
+                variables = comp.values() if comp.is_indexed() else (comp,)
+                _filter_model_variables(
+                    variable_set,
+                    variables,
+                    include_continuous,
+                    include_binary,
+                    include_integer,
+                    include_fixed,
+                )
```
I'm not sure.  @jlgearh ???

emma58 replied:
We discussed this and this is here because it is allowed to pass in Vars to this method and this acts as a sort of flattener. This code could be simplified, but after this PR.

whart222 commented on pyomo/contrib/alternative_solutions/aos_utils.py:
```diff
@@ -0,0 +1,290 @@
+                    include_fixed,
+                )
+            else:  # pragma: no cover
+                if not quiet:
+                    print(
+                        ("No variables added for unrecognized component {}.").format(
+                            comp
+                        )
+                    )
```
In principle, I'm OK with this.  But I don't know how to use the Pyomo logger.  Is it as easy as passing-in a boolean flag?

I'm inclined to make this change later, since I wouldn't know how to implement or test it.

emma58 replied:
It's pretty easy--mostly a find and replace of print statements, though you'll have to choose the logging level for them. We usually just do something along the lines of:
```
import logging

logger = logging.getLogger(__name__)
```
at the top of the file. 

Then, instead of print statements, you can just do: `logger.warn("No variables added...")` or `logger.debug("Starting No-Good cut analysis")`, etc. Then users can just set the logging level to what they want, direct it to a file instead of stdout, or whatever, and you don't have to handle any of that logic. The levels and things are in [these docs](https://docs.python.org/3/howto/logging.html)

And if you want to test it, you can use `pyomo.common.log.LoggingIntercept`, but I think that can definitely wait until after this PR for that.

whart222 replied:
Done

whart222 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,246 @@
+        solver : string
+            The solver to be used.
+        solver_options : dict
+            Solver option-value pairs to be passed to the solver.
```
Maybe.  But under the hood we're using two different types of solvers, which are configured differently depending on the solver interface.

There's a bigger question of how to consolidate the solver interfaces used in this PR, now that they appear to have been consolidated within Pyomo.  But that's definitely a post-PR change.

whart222 replied:
I'm going to mark this as resolved this b.c. this is a post-PR change.

whart222 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,246 @@
+            A list of Solution objects.
+            [Solution]
+    """
+    if not quiet:  # pragma: no cover
+        print("STARTING NO-GOOD CUT ANALYSIS")
```
See above.

jsiirola replied:
The easiest way to use the logger is to (at module scope):
```python
import logging
logger = logging.getLogger(__name__)
```
Then when you want to output text, instead of `print(msg)`, you would just do `logger.info(msg)` or `logger.debug(msg)`.  You can log at any integer level, but we tend to stick to the standard 5 (debug, info, warning, error, critical).

whart222 replied:
Done

whart222 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,257 @@
+    if not quiet:  # pragma: no cover
+        print("Performing initial solve of model.")
+    results = opt.solve(model, tee=tee, load_solutions=False)
+    status = results.solver.status
+    condition = results.solver.termination_condition
```
Done.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,354 @@
+        Solver option-value pairs to be passed to the solver.
+    tee : boolean
+        Boolean indicating that the solver output should be displayed.
+    quiet : boolean
+        Boolean indicating whether to suppress all output.
+    debug : boolean
+        Boolean indicating whether to include debugging output.
```
See above

whart222 replied:
Done

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,354 @@
+    if not quiet:  # pragma: no cover
+        print("STARTING LP ENUMERATION ANALYSIS")
+
+    # TODO: Make this a parameter
+    zero_threshold = 1e-5
```
It's not clear that we *should* make this a parameter for the public API.  I've clarified that in the comment.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,354 @@
+        "random",
+        "norm",
+    ], 'search mode must be "optimal", "random", or "norm".'
+    # TODO: Implement the random and norm objectives. I think it is sufficient
+    # to only consider the cb.var_lower variables in the objective for these two
+    # cases. The cb.var_upper variables are directly linked to these to diversity
+    # in one implies diversity in the other. Diversity in the cb.basic_slack
+    # variables doesn't really matter since we only really care about diversity
+    # in the original problem and not in the slack space (I think)
```
Not sure.  @jlgearh ???

emma58 replied:
This can be resolved after the PR also. It is correct but greedy as written.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum.py:
```diff
@@ -0,0 +1,354 @@
+
+    # For now keeping things simple
+    # TODO: See if this can be relaxed, but for now just leave as all
+    assert variables == "all"
```
I want to keep this.  We intend to generalize this functionality "soon", so establishing the API now seems appropriate.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,234 @@
+            A list of Solution objects.
+            [Solution]
+    """
+    print("STARTING LP ENUMERATION ANALYSIS USING GUROBI SOLUTION POOL")
```
Again, so my note about logging.

This comment applies to *a lot* of output in these methods.

whart222 replied:
Done

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,234 @@
+    # Setup gurobi
+    #
+    if not gurobi_available:
+        return []
```
I don't see the risk.  The solver is not available, then no solutions are returned.  Other methods in this package behave similarly.

Exception management is painful, and it seems worthwhile to avoid exposing this pain to end-users.

whart222 replied:
Raising exceptions, since that seems to be the way other contrib packages are handling this error.

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,234 @@
+
+    orig_objective = aos_utils.get_active_objective(model)
+    orig_objective_value = pe.value(orig_objective)
+    print("Found optimal solution, value = {}.".format(orig_objective_value))
```
See my comments above.

whart222 replied:
Done

whart222 commented on doc/OnlineDocs/contributed_packages/alternative_solutions.rst:
```diff
@@ -0,0 +1,102 @@
+
+* alternative solutions can support an assessment of trade-offs between competing objectives;
+
+* if the optimization formulation may be inaccurate or untrustworthy, then comparisons amongst alternative solutions provides additional insights into the reliability of these model predictions; or
```
Done

whart222 commented on doc/OnlineDocs/contributed_packages/alternative_solutions.rst:
```diff
@@ -0,0 +1,102 @@
+
+* ``enumerate_linear_solutions``
+
+    * Finds alternative optimal solutions a (mixed-integer) linear program.
```
Done

whart222 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,370 @@
+    warmstart : boolean
+        Boolean indicating that the solver should be warmstarted from the
+        best previously discovered solution.
+    solver : string
+        The solver to be used.
+    solver_options : dict
+        Solver option-value pairs to be passed to the solver.
```
See my comments above

whart222 commented on pyomo/contrib/alternative_solutions/obbt.py:
```diff
@@ -0,0 +1,370 @@
+        Solver option-value pairs to be passed to the solver.
+    tee : boolean
+        Boolean indicating that the solver output should be displayed.
+    quiet : boolean
+        Boolean indicating whether to suppress all output.
```
See my comments above

whart222 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,257 @@
+        condition = results.solver.termination_condition
+        if pe.check_optimal_termination(results):
+            model.solutions.load_from(results)
+            orig_obj_value = pe.value(orig_objective)
+            orig_obj_value = pe.value(orig_objective)
```
Fixed

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,234 @@
+            self.solutions.append(sol)
+
+            if len(self.solutions) >= self.num_solutions:
+                # TODO: (nicely) terminate the solve
+                return
```
@jlgearh ???

whart222 commented on pyomo/contrib/alternative_solutions/lp_enum_solnpool.py:
```diff
@@ -0,0 +1,234 @@
+    if variables == "all":
+        all_variables = aos_utils.get_model_variables(model, "all")
+
+    # TODO: Check if problem is continuous or mixed binary
```
@jlgearh ???

emma58 replied:
This needs to throw an error if there are binaries. It's only implemented for continuous now.

whart222 replied:
Done

emma58 replied:
Unresolved this for the moment because I talked with @jlgearh and this needs to check that no variables are discrete to be correct--he plans to make the change.

jlgearh replied:
I updated this to check for any discrete variables.

whart222 commented on doc/OnlineDocs/contributed_packages/alternative_solutions.rst:
```diff
@@ -0,0 +1,102 @@
+Usage Example
+-------------
+
+Many of functions in the alternative-solutions library have similar options, so we simply illustrate the ``enumerate_binary_solutions`` function.  We define a simple model whose feasible space is a isosceles right triangle. The optimal solutiosn fall along the hypotenuse, where :math:`x + y == 5`.  Alternative near-optimal feasible points have integer objective values ranging from 0 to 4.
```
Done

whart222 commented on pyomo/contrib/alternative_solutions/balas.py:
```diff
@@ -0,0 +1,246 @@
+                )
+                print(", ".join(non_binary_variables))
+
+    all_variables = aos_utils.get_model_variables(model, "all", include_fixed=True)
```
Done

whart222 commented:
Tests seem to be working, except for a conda segfault

emma58 commented:
@jlgearh, @whart222, I just opened #3342 to track planned changes not addressed in this PR. Please make sure I didn't miss anything when you get a chance.

whart222 commented on doc/OnlineDocs/contributed_packages/alternative_solutions.rst:
```diff
@@ -0,0 +1,107 @@
+###############################################
+
+Optimization solvers are generally designed to return a feasible solution
+to the user. However, there are many applications where a users needs
```
Done

whart222 commented on doc/OnlineDocs/contributed_packages/alternative_solutions.rst:
```diff
@@ -0,0 +1,107 @@
+
+* ``gurobi_generate_solutions``
+
+    * Finds alternative optimal solutions for discrete variables using Gurobi's built-in Solution Pool capability.
```
Done
